src/manifest.c

```diff
@@ -1391,3 +1476,73 @@ struct file *search_file_in_manifest(struct manifest *manifest, const char *file
 	}
 	return NULL;
 }
+
+/* Ideally have the manifest sorted already by this point */
+struct file **manifest_files_to_array(struct manifest *manifest)
+{
+	struct file **array;
+	struct list *iter;
+	struct file *file;
+	int i = 0;
+	int numfiles = manifest->filecount;
+
+	if (!manifest) {
+		fprintf(stderr, "Manifest is NULL!\n");
+		return NULL;
+	}
+
+	array = malloc(sizeof(struct file *) * numfiles);
+
+	iter = list_head(manifest->files);
+	while (iter) {
+		file = iter->data;
+		iter = iter->next;
+		array[i] = file;
+		i++;
+	}
+	return array;
+}
+
+void print_manifest_array(struct file **array, int filecount)
+{
+	for (int i = 0; i < filecount; i++) {
+		printf("[%d]/%d %s\n", i, filecount, array[i]->filename);
+	}
+	printf("SIZE: %d\n", filecount);
+}
+
+void free_manifest_array(struct file **array)
+{
+	free(array);
+}
+
+static int cmpnames(const void *a, const void *b)
+{
+	return strcmp((*(struct file **)a)->filename, (*(struct file **)b)->filename);
+}
+
+int enforce_compliant_manifest(struct file **a, struct file **b, int searchsize, int size)
+{
+	struct file **found;
+	int ret = 0;
+
+	qsort(a, searchsize, sizeof(struct file *), cmpnames);
```

Comment:
It won't work with the way it's sorted to begin with. The sort function in swupd sorts based on multiple heuristics and bsearch does not play well with that because it's not _really_ sorted.

Reply:
The code says

sort the array 'a', using the filename as the key
sort the array 'b', using the filename as the key.
loop through the elements of 'a' in order, searching for each of them in 'b'.
     search, using bsearch, the 'b' array for this element.
     if found, then ensure that the hash is the same....
There is nothing here that seems to depend on processing the elements in 'a' in alphabetical order. You would get the same return value if you processed them in reverse order, or a random order. Hence my suggestion to not bother sorting 'a'.

Now if you choosing to step through the two arrays at the same time, rather than stepping through 'a' and searching for matching elements in 'b' with bsearch, you would want both sorted. Your costs would then be n.log(n) for each of the 2 sorts, and n for the compare, rather than n.log(n) for the sort of 'b' and n (length of a) log(n) (bsearch of b).